File: trainer.lua
require 'nn'
require 'sys'
require 'torch'

require 'util'

local Trainer = torch.class('Trainer')

-- Perform one epoch of training.
function Trainer:train(train_data, train_labels, model, criterion, optim_method, layers, state, params, grads, opt)
  model:training()

  local train_size = train_data:size(1)

  local timer = torch.Timer()
  local time = timer:time().real
  local total_err = 0

  local classes = {}
  for i = 1, opt.num_classes do
    table.insert(classes, i)
  end
  local confusion = optim.ConfusionMatrix(classes)
  confusion:zero()

  local config -- for optim
  if opt.optim_method == 'adadelta' then
    config = { rho = 0.95, eps = 1e-6 } 
  elseif opt.optim_method == 'adam' then
    config = {}
  end

  -- shuffle batches
  local num_batches = math.floor(train_size / opt.batch_size)
  local shuffle = torch.randperm(num_batches)
  for i = 1, shuffle:size(1) do
    local t = (shuffle[i] - 1) * opt.batch_size + 1
    local batch_size = math.min(opt.batch_size, train_size - t + 1)

    -- data samples and labels, in mini batches.
    local inputs = train_data:narrow(1, t, batch_size)
    local targets = train_labels:narrow(1, t, batch_size)
    if opt.cudnn == 1 then
      inputs = inputs:cuda()
      targets = targets:cuda()
    else
      inputs = inputs:double()
      targets = targets:double()
    end

    -- closure to return err, df/dx
    local func = function(x)
      -- get new parameters
      if x ~= params then
        params:copy(x)
      end
      -- reset gradients
      grads:zero()

      -- forward pass
      local outputs = model:forward(inputs)
      local err = criterion:forward(outputs, targets)

      -- track errors and confusion
      total_err = total_err + err * batch_size
      for j = 1, batch_size do
        confusion:add(outputs[j], targets[j])
      end

      -- compute gradients
      local df_do = criterion:backward(outputs, targets)
      model:backward(inputs, df_do)

      if opt.model_type == 'static' then
        -- don't update embeddings for static model
        layers.w2v.gradWeight:zero()
      elseif opt.model_type == 'multichannel' then
        -- keep one embedding channel static
        layers.chan1.gradWeight:zero()
      end

      return err, grads
    end

    -- gradient descent
    optim_method(func, params, config, state)
    -- reset padding embedding to zero
    layers.w2v.weight[1]:zero()
    if opt.model_type == 'multichannel' then
      layers.chan1.weight[1]:zero()
    end
    if opt.skip_kernel > 0 then
      -- keep skip kernel at zero
      layers.skip_conv.weight:select(3,3):zero()
    end

    -- Renorm (Euclidean projection to L2 ball)
    local renorm = function(row)
      local n = row:norm()
      row:mul(opt.L2s):div(1e-7 + n)
    end

    -- renormalize linear row weights
    local w = layers.linear.weight
    for j = 1, w:size(1) do
      renorm(w[j])
    end
  end

  if opt.debug == 1 then
    print('Total err: ' .. total_err / train_size)
    print(confusion)
  end

  -- time taken
  time = timer:time().real - time
  time = opt.batch_size * time / train_size
  if opt.debug == 1 then
    print("==> time to learn 1 batch = " .. (time*1000) .. 'ms')
  end

  -- return error percent
  confusion:updateValids()
  return confusion.totalValid
end

function Trainer:test(test_data, test_labels, model, criterion, layers, dump_features, opt)
  model:evaluate()

  local classes = {}
  for i = 1, opt.num_classes do
    table.insert(classes, i)
  end
  local confusion = optim.ConfusionMatrix(classes)
  confusion:zero()

  local preds_file
  if opt.test_only == 1 and opt.preds_file ~= '' then
    print('Writing predictions to ' .. opt.preds_file)
    preds_file = io.open(opt.preds_file, 'w')
  end

  -- dump feature maps
  local feature_maps
  local conv_layer = get_layer(model, 'convolution')

  local test_size = test_data:size(1)
  local total_err = 0
  for t = 1, test_size, opt.batch_size do
    -- data samples and labels, in mini batches.
    local batch_size = math.min(opt.batch_size, test_size - t + 1)
    local inputs = test_data:narrow(1, t, batch_size)
    local targets = test_labels:narrow(1, t, batch_size)
    if opt.cudnn == 1 then
      inputs = inputs:cuda()
      targets = targets:cuda()
    else
      inputs = inputs:double()
      targets = targets:double()
    end

    local outputs = model:forward(inputs)
    -- dump feature maps from model forward
    local cur_feature_maps
    if dump_features then
      if opt.cudnn == 1 then
        cur_feature_maps = conv_layer.output:squeeze(4)
      else
        cur_feature_maps = conv_layer.output
      end
      if feature_maps == nil then
        feature_maps = cur_feature_maps
      else
        feature_maps = torch.cat(feature_maps, cur_feature_maps, 1)
      end
    end

    if opt.test_only == 1 and opt.preds_file ~= '' then
      -- write predictions to file
      local _,preds = torch.max(outputs, 2)
      for j = 1, preds:size(1) do
        -- zero index
        preds_file:write((preds[j][1] - 1) .. '\n')
      end
    end

    local err = criterion:forward(outputs, targets)
    total_err = total_err + err * batch_size

    for i = 1, batch_size do
      confusion:add(outputs[i], targets[i])
    end
  end

  if opt.debug == 1 then
    print(confusion)
    print('Total err: ' .. total_err / test_size)
  end

  if dump_features then
    assert(#opt.kernels == 1, 'multiple kernels not yet supported')
    local kernel = opt.kernels[1]
    local word_idxs = test_data:narrow(2, kernel, test_data:size(2) - kernel + 1)
    assert(feature_maps:size(1) == word_idxs:size(1), 'length mismatch')

    local filename = opt.dump_feature_maps_file .. '.hdf5'
    print('dumping features to ' .. filename)
    local f = hdf5.open(filename, 'w')
    f:write('feature_maps', feature_maps:double())
    f:write('word_idxs', word_idxs:long())
    f:close()
  end

  if opt.test_only == 1 and opt.preds_file ~= '' then
    preds_file:close()
  end

  -- return error percent
  confusion:updateValids()
  return confusion.totalValid
end

return Trainer

File: trainer.py
"""
Convolutional net for sentence classification
"""

import torch
import torch.nn as nn
import torch.nn.functional as F
import argparse
import torch.optim as optim
import numpy as np

parser = argparse.ArgumentParser(
    description=__doc__,
    formatter_class=argparse.RawDescriptionHelpFormatter)
parser.add_argument('infile', help="Input file", 
                    type=argparse.FileType('r'))
parser.add_argument('-o', '--outfile', help="Output file",
                    default=sys.stdout, type=argparse.FileType('w'))

args = parser.parse_args(arguments)

parser.add_argument('data', default='', help='Training data and word2vec data')
parser.add_argument('--model_type', default='nonstatic',
                    help='Model type. Options: rand (randomly initialized word embeddings), static (pre-trained embeddings from word2vec, static during learning), nonstatic (pre-trained embeddings, tuned during learning), multichannel (two embedding channels, one static and one nonstatic)')

def cmd_option(name, default, help_str):
    parser.add_argument("-" + name, default=default, help=help_str)

# Training hyperparameters
cmd_option('-num_epochs', 25, 'Number of training epochs')
cmd_option('-optim_method', 'adadelta', 'Gradient descent method. Options: adadelta, adam')
cmd_option('-L2s', 3, 'L2 normalize weights')
cmd_option('-batch_size', 50, 'Batch size for training')

# Model hyperparameters
cmd_option('-num_feat_maps', 100, 'Number of feature maps after 1st convolution')
cmd_option('-kernels', '[3,4,5]', 'Kernel sizes of convolutions, table format.')
cmd_option('-skip_kernel', 0, 'Use skip kernel')
cmd_option('-dropout_p', 0.5, 'p for dropout')
cmd_option('-highway_mlp', 0, 'Number of highway MLP layers')
cmd_option('-highway_conv_layers', 0, 'Number of highway MLP layers')


# cmd_option('-cudnn', 0, 'Use cudnn and GPUs if set to 1, otherwise set to 0')
# cmd_option('-seed', 3435, 'random seed, set -1 for actual random')
# cmd_option('-folds', 10, 'number of folds to use. If test set provided, folds=1. max 10')
# cmd_option('-debug', 0, 'print debugging info including timing, confusions')
# cmd_option('-gpuid', 0, 'GPU device id to use.')
# cmd_option('-savefile', '', 'Name of output file, which will hold the trained model, model parameters, and training scores. Default filename is TIMESTAMP_results')
# cmd_option('-zero_indexing', 0, 'If data is zero indexed')
# cmd_option('-dump_feature_maps_file', '', 'Set file to dump feature maps of convolution')
# cmd:text()

# -- Preset by preprocessed data
# cmd_option('-has_test', 1, 'If data has test, we use it. Otherwise, we use CV on folds')
# cmd_option('-has_dev', 1, 'If data has dev, we use it, otherwise we split from train')
# cmd_option('-num_classes', 2, 'Number of output classes')
# cmd_option('-max_sent', 59, 'maximum sentence length')
# cmd_option('-vec_size', 300, 'word2vec vector size')
# cmd_option('-vocab_size', 18766, 'Vocab size')
# cmd:text()

# -- Training own dataset
# cmd_option('-train_only', 0, 'Set to 1 to only train on data. Default is cross-validation')
# cmd_option('-test_only', 0, 'Set to 1 to only do testing. Must have a -warm_start_model')
# cmd_option('-preds_file', '', 'On test data, write predictions to an output file. Set test_only to 1 to use')
# cmd_option('-warm_start_model', '', 'Path to .t7 file with pre-trained model. Should contain a table with key \'model\'')
# cmd:text()

def train_epoch(net, inputs, targets):
    optimizer = optim.SGD(net.parameters(), lr = 0.01)

    for i in range(0, inputs.size()[0], 50):
        input = inputs[i:i+50]
        target = targets[i:i+50]
        optimizer.zero_grad() 
        output = net(input)
        loss = F.cross_entropy(output, target)
        loss.backward()
        optimizer.step()

class ConvNet(nn.Module):
    def __init__(self, opt):
        super(Net, self).__init__()
        self.word_lut = nn.LookupTable(opt.vocab_size, opt.vec_size)
        self.conv = [nn.TemporalConvolution(opt.vec_size, opt.num_feat_maps, opt.kernels[i])
                     for k in opt.kernels] 
        self.linear = nn.Linear(opt.num_feat_maps * len(opt.kernels), opt.num_classes)
        self.dropout = nn.Dropout(opt.dropout_p)
        
    def forward(self, x):
        x = self.word_lut(x)
        x = torch.cat([F.relu(self.conv[k](x)).data.max(2)
                       for k in opt.kernels], 2)
        x = self.linear(self.dropout(x))
        return F.log_softmax(x)
    
def main(arguments):
    opt = parser.parse_args(arguments)
    
    with h5py.File(opt.data) as f:
        train = np.array(f["train"])
        train_label = np.array(f["train_label"])

    opt.data()
    net = ConvNet(opt)
    train_epoch(train, train_label, net)
    


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
